File: Data Structure/Array Or Vector/Next Greater Element/SolutionbyBharat.cpp
#include<bits/stdc++.h>
#define ll long long int

using namespace std;
typedef vector<ll>vi;


void Solve(vi &array, vi & greater_element, ll array_size){

    ll i=array_size-2;

    while(i>=0){
        if(array[i]<array[i+1])
            greater_element[i]=array[i+1];
        
        ll j=i;
        while(j<array_size){
            if(array[i]<greater_element[j]){
                greater_element[i]=greater_element[j];
                break;   
            }
            j++;
        }

        i--;    
    }

}

int main(){
    ll n;
    cin>>n;

    vi array(n),greater_element(n,-1);

    for(int i=0;i<n;i++){
        cin>>array[i];
    }

    Solve(array,greater_element,n);  

    cout<<"Element"<<" "<<"NGE"<<endl;

    for(ll i=0;i<n;i++)
        cout<<array[i]<<" ----> "<<greater_element[i]<<endl;

   return 0;
}

File: Data Structure/Array Or Vector/Next Greater Element/SolutionbyBharat.py
'''
Problem Statement:-Given an array, print the Next Greater Element (NGE) for every element. The Next greater Element for an element x is the first greater element on the right side of x in array. Elements for which no greater element exist, consider next greater element as -1.

Time Complexity: O(n)
Space Complexity: O(n)

'''

n=int(input())

arr=[int(i) for i in input().split()]
greater_element=[int(-1) for i in range(n)]


i=len(arr)-2

while i>=0:
    if arr[i]<arr[i+1]:
        greater_element[i]=arr[i+1]
    elif arr[i]<greater_element[i+1]:
        greater_element[i]=greater_element[i+1]

    i-=1


print("Element"+"  "+"NGE")

for i in range(n):
    print(arr[i],"--->",greater_element[i])
